File: C++/longest_repeating_character_replacement.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        int res = 0; int l = 0;
        unordered_map<char, int> window;

        for (int r = 0; r < s.size(); r++){
            if (!window.contains(s[r])){
                window[s[r]] = 0;
            }
            window[s[r]]++;
            int maxVal = INT_MIN;
            for (const auto& [k, v] : window){maxVal = max(maxVal, v);}
            
            while (maxVal + k < r - l + 1){
                window[s[l]]--;
                l++;
            }

            res = max(res, r - l + 1);
        }

        return res;
    }
};

File: python/longest_repeating_character_replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        res = 0
        window = defaultdict(int)
        l = 0

        for r in range(len(s)):
            window[s[r]] += 1
            while max(window.values()) + k < r - l + 1:
                window[s[l]] -= 1
                l += 1
            res = max(res, r - l + 1)
        return res


